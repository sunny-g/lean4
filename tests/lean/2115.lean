class A (α : Type) where
  one : α

class B (α : Type) extends A α where
  two : α

class C (α : Type) extends A α where
  three : α

class D (α : Type) extends B α, C α

attribute [instance 10] C.toA

set_option pp.all true

def foo [D α] : A α := inferInstance -- should use `B.toA` since it has higher priority
#print foo

attribute [instance 5] B.toA

def bla [D α] : A α := inferInstance -- should use `C.toA` since it has higher priority
#print bla

attribute [-instance] C.toA

def boo [D α] : A α := inferInstance -- should use `B.toA`, `C.toA` attribute was erased
#print boo

attribute [instance 10] C.toA

def boo2 [D α] : A α := inferInstance -- should use `C.toA` since it has higher priority
#print boo2
