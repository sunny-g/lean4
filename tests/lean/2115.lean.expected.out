def foo : {α : Type} → [inst : D α] → A α :=
fun {α : Type} [inst : D α] => @inferInstance.{1} (A α) (@B.toA α (@D.toB α inst))
def bla : {α : Type} → [inst : D α] → A α :=
fun {α : Type} [inst : D α] => @inferInstance.{1} (A α) (@C.toA α (@D.toC α inst))
def boo : {α : Type} → [inst : D α] → A α :=
fun {α : Type} [inst : D α] => @inferInstance.{1} (A α) (@B.toA α (@D.toB α inst))
def boo2 : {α : Type} → [inst : D α] → A α :=
fun {α : Type} [inst : D α] => @inferInstance.{1} (A α) (@C.toA α (@D.toC α inst))
